### ISSUE TITLE
It seems does't support arm Mac

### ISSUE THREAD

**Issue by 251305060** [rgov/Thunderbolt3Unblocker#62]
[t3u-diagnose.txt](https://github.com/rgov/Thunderbolt3Unblocker/files/5578040/t3u-diagnose.txt)
I try it on MacBook Air m1, but it looks that the code does't support arm deviceExecuting: /usr/bin/kmutil install --volume-root / --check-rebuild
Not rebuilding KC for arch: 'arm64e'
kmutil done
Executing: /usr/bin/kmutil load -b es.govost.ryan.Thunderbolt3Unblocker
Incompatible architecture: Binary is for x86_64, but needed arch arm64e

**Comment by rgov** [comment]
I don't know if it is possible to support the M1 Macs. Do they support third-party kernel extensions? Apple has been moving to eliminate them for several releases now.

**Comment by 251305060** [comment]
If third party extensions is blocked, is it have other possibility to unblock tb3. such as add md5 string of 11.0.1 to TB3 enabler
5

**Comment by rgov** [comment]
Other methods of unblocking involve patching the kernel extensions on disk. I don't know if those methods still work. They might, but Apple is continuously adding more protections to the system. You should probably look into getting a supported dock.

**Comment by qinyongqiang** [comment]
> Other methods of unblocking involve patching the kernel extensions on disk. I don't know if those methods still work. They might, but Apple is continuously adding more protections to the system. You should probably look into getting a supported dock.

I cloned the project and try to compile it in universal mode (support both arm and intel) and it seems that some asm instructions not working, see attached screenshot.
<img width="1122" alt="截屏2020-12-07 15 22 11" src="https://user-images.githubusercontent.com/23164059/101321465-5fe99b00-38a0-11eb-8372-1757846d9dc7.png">

**Comment by rgov** [comment]
Ah yeah. It does need to be updated for ARM. I do not have an M1 Mac so I am unsure what changes need to be made.

**Comment by qinyongqiang** [comment]
> Ah yeah. It does need to be updated for ARM. I do not have an M1 Mac so I am unsure what changes need to be made.

Mine is on the way...It seems that the __asm__ mov cr0 or cr3 is only available in Intel. I'm not sure if this instruction could be translated by Rosetta

**Comment by rgov** [comment]
Kernel extensions would not be able to run under Rosetta. (I am still not sure this can be loaded at all.)

Basically the whole `xnu_override.c` would need to be re-implemented for ARM.

**Comment by qinyongqiang** [comment]
> Kernel extensions would not be able to run under Rosetta. (I am still not sure this can be loaded at all.)
> 
> Basically the whole `xnu_override.c` would need to be re-implemented for ARM.

It seems that would be quite a lot of working... How can I test if the kext can be loaded in M1?

**Comment by Nevin8618** [comment]
HP ZBook ThunderBolt Dock can directily use under M1 MacBook Air without applying this kernel patch, So if you have a M1 chip Mac, you can try plug the dock without patching this driver.

**Comment by qinyongqiang** [comment]
> HP ZBook ThunderBolt Dock can directily use under M1 MacBook Air without applying this kernel patch, So if you have a M1 chip Mac, you can try plug the dock without patching this driver.

Really? It's a good news. I just got my M1 Air but haven't got a chance to test. I'll have a test and then post updates here.

**Comment by Nevin8618** [comment]
> > HP ZBook ThunderBolt Dock can directily use under M1 MacBook Air without applying this kernel patch, So if you have a M1 chip Mac, you can try plug the dock without patching this driver.
> 
> Really? It's a good news. I just got my M1 Air but haven't got a chance to test. I'll have a test and then post updates here.

I‘ve just tested when I got my M1 Air, and I asked a seller who have this dock on sell. He comfirmed that the dock is perfectly suppoted. All interfaces are perfectly worked. And displayport can output 4k@60Hz after the dock updated the letest firmware.

**Comment by kosztag** [comment]
Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?

**Comment by Nevin8618** [comment]
> Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?

Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.

**Comment by kosztag** [comment]
> > Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?
> 
> Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.

Ok, but I already have Dell TB16 dock using with my Dell laptop. It would be fine if I can us the same dock for both.

**Comment by Nevin8618** [comment]
> > > Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?
> > 
> > 
> > Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.
> 
> Ok, but I already have Dell TB16 dock using with my Dell laptop. It would be fine if I can us the same dock for both.

TB16 Dock has problems in PD charging. So the problem may not related to the driver or that saying "apple's blocking".

**Comment by qinyongqiang** [comment]
> > Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?
> 
> Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.

Do you mean this dock? https://support.hp.com/gb-en/document/c04959218
Interesting...Intel Macs doesn't support this dock yet and patching is must.

**Comment by Nevin8618** [comment]
> > > Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?
> > 
> > 
> > Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.
> 
> Do you mean this dock? https://support.hp.com/gb-en/document/c04959218
> Interesting...Intel Macs doesn't support this dock yet and patching is must.

Yes, surely this one. Intel Macs can't support this but M1 Macs can use this without any patching.

**Comment by dale19900108** [comment]
> > > Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?
> > 
> > 
> > Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.
> 
> Do you mean this dock? https://support.hp.com/gb-en/document/c04959218
> Interesting...Intel Macs doesn't support this dock yet and patching is must.

I was so confused about  this dock  ，it work well on my Intel Macs....M1  ，I'm not sure

**Comment by baronmax** [comment]
Yes I tried the HP ZBook / Elitebook TB3 dock with Macbook M1 and it seemed to work despite showing as 'unsupported'.  I tested the USB connection only as I only have one TB3 cable at present.

**Comment by gilesdavison** [comment]
> > > Hi, I'm also in this situation. I have MacBook Pro M1, and if you doesn't mind, I would like to test. Could someone share me the fixed code or binary?
> > 
> > 
> > Macs with M1 chip support HP ZBook ThunderBolt Dock natively, and no need for any kernel patch.
> 
> Ok, but I already have Dell TB16 dock using with my Dell laptop. It would be fine if I can us the same dock for both.

I'm guessing the M1 doesn't support TB16 natively, did you resolve this?

**Comment by General13K7** [comment]
HP ZBook 150W Thunderbolt 3 Dock  this dock is no longer for sale.
So no testing with thunderbolt unlock?

**Comment by gencymex** [comment]
I think that I found out what we need to do to get this running on ARM Macs. According to the [Apple Support doc](https://support.apple.com/guide/deployment-reference-macos/kernel-extensions-in-macos-apd37565d329/web) it is possible to run kernel extensions, but not recommended for security purposes. 

Once you lower your firmware security level in RecoveryOS, compile the project with Xcode. I haven't taken a look at the xcode project yet, as I have just discovered this awesome repo and wanted to look into the possibility first.

**Comment by rgov** [comment]
There is a source file that is highly specific to Intel processors that has no hope of working on ARM without a developer porting it. Simply using Xcode is not sufficient.

**Comment by tanpengsccd** [comment]
> the

YES ,It work on M1 MBP macOS 11.2.3,but when I unplug my External hard disk from this HP TB3 dock , my MBP  Crash and reboot. Are you having this problem?

this Dock:
https://support.hp.com/us-en/drivers/selfservice/hp-zbook-dock-with-thunderbolt-3/11122586
